### PR TITLE
Test disabling HTTP caching

### PR DESCRIPTION
**Pull request by JunielKatarn**
## Description

~~Explicitly disables The HTTP module's client object caching capability.~~

### Why
There are scenarios where an HTTP endpoint may return different content for the exact same GET request (URL, headers).
This behavior isn't currently validated in RNW's automation.

### What
Test the HTTP client cache behavior when the server returns different content for identical requests, setting the `Cache-Control: no-cache` header.

### Testing
Defined integration test `GetDisableThenReenableCacheSucceeds`.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/12103)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Heads up that this behavior is different from React Native core. At least on Android, we keep a 10MB request cache, and respect RFC 7234 behavior. 

1. https://github.com/facebook/react-native/blob/517fee6dd16e8f399b4c89b92df347bb1cf6efda/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/OkHttpClientProvider.java
2. https://square.github.io/okhttp/features/caching/

File: vnext/Shared/Networking/RedirectHttpFilter.cpp [hunk 1]
```diff
@@ -50,10 +52,14 @@ RedirectHttpFilter::RedirectHttpFilter(
   if (auto baseFilter = m_innerFilter.try_as<IHttpBaseProtocolFilter>()) {
     baseFilter.AllowAutoRedirect(false);
     baseFilter.AllowUI(false);
+    baseFilter.CacheControl().ReadBehavior(HttpCacheReadBehavior::NoCache);
+    baseFilter.CacheControl().WriteBehavior(HttpCacheWriteBehavior::NoCache);
   }
   if (auto baseFilter = m_innerFilterWithNoCredentials.try_as<IHttpBaseProtocolFilter>()) {
     baseFilter.AllowAutoRedirect(false);
```
Comment (NickGerleman) [inline vnext/Shared/Networking/RedirectHttpFilter.cpp:59]: RN on Android by default follows browser behavior, and does allow redirects.

Comment (JunielKatarn) [reply vnext/Shared/Networking/RedirectHttpFilter.cpp:59]: See class `RedirectHttpFilter`.
Redirects don't work on some cases deemed as "irregular redirects" by the underlying HTTP stack (WinInet), irrecoverably throwing `ERROR_HTTP_REDIRECT_NEEDS_CONFIRMATION` when invoked from the WinRT layer.

Redirects are still followed, only manually.

**Comment by JunielKatarn** [comment]
> Heads up that this behavior is different from React Native core (and web browsers). At least on Android, we keep a 10MB request cache, and respect RFC 7234 behavior.
> 
> 1. https://github.com/facebook/react-native/blob/517fee6dd16e8f399b4c89b92df347bb1cf6efda/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/OkHttpClientProvider.java
> 2. https://square.github.io/okhttp/features/caching/
> 
> I suspect disabling cache for every request could regress anyone fetching static assets.
> 
> For this specific request, does RNW caching behavior conform to server cache headers, and does WinRT network stack respect cache control client headers? If so, it seems like either client code making this specific request, or server code, need to request different caching behavior.

We'll evaluate making caching configurable.

**Comment by rozele** [comment]
Does adding a Cache-Control: no-cache header work? Or could we use that to configure per request?

**Comment by JunielKatarn** [comment]
> Does adding a Cache-Control: no-cache header work? Or could we use that to configure per request?

It does, but it had to be done on the server side.
I am reverting the changes to the HTTP module code and adding an integration test to validate the use case.

**Review by jonthysell — APPROVED** [review]

**Pull request merged**